packages/core/src/utils.ts

```diff
@@ -328,30 +324,36 @@ export function isString(value: any): value is string {
 }
 
 export function toGuard<TContext, TEvent extends EventObject>(
-  condition?: Condition<TContext, TEvent>,
-  guardMap?: Record<string, ConditionPredicate<TContext, TEvent>>
-): Guard<TContext, TEvent> | undefined {
-  if (!condition) {
-    return undefined;
-  }
-
-  if (isString(condition)) {
+  guardConfig: GuardConfig<TContext, TEvent>,
+  guardMap?: Record<string, GuardPredicate<TContext, TEvent>>
```

Comment:
Not a specific comment for this PR so not a blocker at all - but I find it confusing for a function like this (there are other similar ones - e.g. when creating action objects) to be used kinda twice. Once when creating action objects without the context of the machine (where `guardMap` is undefined) and the second time when we resolve the previously created guard object with `guardMap` available.

IMHO the overall algorithm would be simpler if this wouldn't be aware of `guardMap` at all and later only some function like `getGuardPredicate` would get the actual predicate from the `guardMap` based on the object's type and just call it with that object - instead of going through `toGuard` again.

Reply:
You're right... let's think about how to do this in another PR for sure.